README.md
# Rabbit Sky - Client Side (Web Assets)

This is the Client Side / Web Assets / Part One of the Rabbit Sky. This repository contains only HTML, CSS, Javascripts and Images.

Since this part only contains assets, you must run a webserver to serve these assets. We recommend to use NGINX, since it's lightweight and very high performance.

Also, Rabbit Sky cannot run on it's own without the server side of the Rabbit Sky. [Click here to see the Server Side / Part Two](https://github.com/rabbitsky-io/rabbitsky-server).

## Overall Requirement
##### Mandatory
- VM / VPS / Dedicated Server, with minimum 1GB of RAM. Linux OS (Ubuntu / CentOS) is preferred.
- Public IP attached on the VM / VPS / Dedicated Server.
- Web Server such as NGINX, Lighttpd or Apache.
- This Client Assets.
- [RabbitSky Server (Go)](https://github.com/rabbitsky-io/rabbitsky-server).

##### Optional
- Domain / Sub Domain. You can use IP, but having a domain is preferred.
- SSL Certificate. Mandatory if you want to use **Twitch Embed**, see Known Problem below for more detail.

## Download

You can download these repository as zip, or simply clone this git to your webserver folder.

```
git clone git@github.com:rabbitsky-io/rabbitsky-server.git
```

Also you need the Rabbit Sky Server, [Click here to see the Server Side / Part Two](https://github.com/rabbitsky-io/rabbitsky-server).

## Installing

Basically, just download this repo, run the server side, edit the configuration file, run a webserver that point the webroot to the assets folder, and you're good to go!

## Web Server?
This client won't run standalone, it need the helps of webserver to serve the files. We recommend to use NGINX, but there's a lot of other web server like Lighttpd, Apache HTTPD, etc.

If you're interested in using NGINX for your web server, please follow [this tutorial on how to install it](https://docs.nginx.com/nginx/admin-guide/installing-nginx/installing-nginx-open-source/).

Also if you want to have one domain for both the client and server side using nginx, use this example:

```
server {
    listen 80;
    server_name demo.rabbitsky.io;

    # Change this if your folder is different with the default
    root /var/www/rabbitsky-web;

    location /channel/ {
        # this is the default port of the server side
        proxy_pass http://127.0.0.1:8080;
    }

    # Web Socket Upgrade
    location = /channel/join {
        # this is the default port of the server side
        proxy_pass http://127.0.0.1:8080;

        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_set_header Host $host;
    }
}
```

## What's Inside?
This client side use fully written in Plain Javascript, [THREE.js](https://threejs.org/), plain CSS, and Plain HTML. We don't want to use any other framework because we want it to have a good performance, even on mobile phone and toaster pc (also our main goal is to have this run on Samsung Smart Fridge!).

Also we used third party script, like WebFont and EmojiPicker. Some third party scripts was edited so it works great on the Rabbit Sky.

## Configuration File

Configuration file is JSON Formatted, located at: `config.json`

You can copy the `config.example.json` to `config.json` and edit it to your own configuration.

#### Configuration Example

```
{
    "embed": {
        "type": "twitch",
        "id": "monstercat",
        "chat": true
    },
    "servers": [
        {
            "name": "Demo Rabbit Sky",
            "host": "demo.rabbitsky.io",
            "secure": true
        }
    ]
}
```

#### Configuration Explanation
| Variable | Type | Description |
| ----- | ---- | ----------- |
| embed.type | string | Type of embed video, acceptable value are only `twitch` for TwitchTV, and `youtube` for YouTube Live.
| embed.id | string | ID of the embed video. For Twitch, you can use the channel name: `twitch.tv/[channelID]`, so the value is the ChannelID. For YouTube Live, you can use the id in the url: `youtube.com/watch?v=[youtubeID]`, so the value is the YouTubeID.
| embed.chat | boolean | Show live chat on the right side of the game. |
| servers | array | Contains array of the server detail. You can have multiple servers without maximum number. |
| servers[].name | string | Name of the server, it shows in Server List menu. |
| servers[].host | string | Domain / Hostname where the server side is running. You can use IP. Ex: `https://demo.rabbitsky.io/` or `wss://demo.rabbitsky.io` so the value is `demo.rabbitsky.io`. **Please remember to set your Server Side origin parameter to this client domain.** |
| servers[].secure | boolean | Server is using SSL Web Server. If you're using https or wss, you should give this value true. If you're using plain http or ws, set it to false. **We recommend you to use HTTPS and WSS due to security and compability. Read Known Problem below.** |

## Anything Easy to Customize?
Yes, for now you easily can change images, emoji and chat filter. If you need more, you have to write some codes. Is it hard? Maybe.

##### Images
You can customize image such as Banner, Logo, and Skybox. Detail:
| Filename | Type | Resolution | Description |
| -------- | ---- | ---------- | ----------- |
| images/og-image.jpg | image/jpeg | 1200 x 627 | For OpenGraph Preview of your website. It works on all social media such as Facebook, Twitter, and Reddit. Cannot be deleted. |
| images/sky/bottom.png | image/png | 1000 x 1000 | Sky Box, Bottom. Should be transparent if you want to change sky color by command. Bigger resolution is better. Cannot be deleted.  |
| images/sky/top.png | image/png |  1000 x 1000 | Sky Box, Top. Should be transparent if you want to change sky color by command. Bigger resolution is better. Cannot be deleted. |
| images/sky/side.png | image/png |  1000 x 1000 | Sky Box, all 4 sides. Should be transparent if you want to change sky color by command. Bigger resolution is better. Cannot be deleted. |
| images/stage/banner-left.png | image/png | 500 x 2560 | Stage Banner Left Side. Can be transparent. **Can be deleted, the object will be gone.** |
| images/stage/banner-right.png | image/png | 500 x 2560 | Stage Banner Right Side. Can be transparent. **Can be deleted, the object will be gone.** |
| images/stage/logo.png | image/png | 9900 x 500 | Stage Logo on the top. Can be transparent. **Can be deleted, the object will be gone.** |
| images/stage/logo-box.png | image/png | 400 x 400 | Spinning box on top of the speaker, for both left and right. Can be transparent. **Can be deleted, the object will be gone.** |

##### Emoji
You can add, remove, and edit Emoji on `extras/emoji.json`. Format is JSON, emoji is on UTF-8 Unicode. You can see the detail from the file.

##### Chat Filter
You can add, remove, and edit banned keywords on `extras/chatfilter.txt`. One word per line.

## I want to edit [...], can I do that?

Yes, you can edit whatever you like, then publish it for your own web. Changing name, logo, even if you want to remove the footer, you can do it. We do not want to limit what you want to do in Rabbit Sky. If you can do it, just do it.

## In Game Chat Command
For in-game chat command, please refer to the [Server Side](https://github.com/rabbitsky-io/rabbitsky-server) ReadMe file.

## Gamepad Controller Support
Rabbit Sky is now Gamepad Enabled, **BUT** we marked it as BETA. That means it will not works 100% well, there will be problem, bug, or browser hang, and also it is because Gamepad API is still in draft.

If you found any bug, let us know by raising issues.

## Known Problem That Cannot Be Fixed
- Twitch only allow secure website (HTTPS) to embed their player. There is no other way to do it except you must install SSL Certificate on your web server. You can generate SSL Cert for free using LetsEncrypt, or you can use third-party that can change your server to HTTPS like Cloudflare.
- YouTube Live Chat not logged in for some browsers, eventho you login-and-out, live chat still treat you as guest. We still don't know why, they don't have official documentation on how embed their live chat. If you encounter this problem, please change your browser. Chrome works best for us.
- When you are using Firefox, using gamepad sometime causing video marked as autoplay and will be played mute, or even not playing at all. To prevent this, you can use mouse for clicking menu. We reported the problem, but there's no timeline yet when it will be fixed. See: [Bugzilla #1679817](https://bugzilla.mozilla.org/show_bug.cgi?id=1679817).

## I am confusion, can you do it for me?
Depends. We can create your own server for some price, but we do not have a big team, so it will take some time. If you're really interested, hit us up. You can see our contact on our website.

## Donate
Liking this and having a spare of money? Consider donating!

[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://paypal.me/wibisaja)

## Contributing

Yes


CHANGELOG.md
# Rabbit Sky Client Changelog

## Version 0.1.7
- Handle size change on rabbit.
- Add light on stage.
- Add admin panel for better experience for changing sky, light, enable fly, change size, and bpm translator for sky and light. Panel will show on /admin command when the password is valid.
- Add Disable Stage Light option on Setting.
- Fix player not updating position if position is not the same as server.
- Change stage opacity for better lighting.
- Change rabbit duck to smaller size, so it's visible when player is ducking.

## Version 0.1.6.5
- This version is compatible with Server v0.1.6.
- Fix mobile control move always to the side, and add dead zone for look control.
- Removed unused gamepad file.

## Version 0.1.6.4
- This version is compatible with Server v0.1.6.
- Change Menu Background Color
- Fix rabbit animation always update, causing some performance issue.
- Improve smooth animation so it is not updating non-visible rabbit on camera.

## Version 0.1.6.3
- This version is compatible with Server v0.1.6.
- Fix escape button showing options in menu.
- Fix chaterror box position after redesign menu.
- Add "Adjust Volume" help on first game helper.
- Change escape button on help.

## Version 0.1.6.2
- This version is compatible with Server v0.1.6.
- Change a bit UI on the Menu Join Button.
- Fix UI Menu on mobile, logo was splitting when view is small.

## Version 0.1.6.1
- This version is compatible with Server v0.1.6
- Fix a bug when rejoin from back to menu button, it will sent an error, but actually it is connected.
- Fix CSS Outline on Link.

## Version 0.1.6
- Full Support of Gamepad Controller (Beta).
- Add `/fly` handler.
- Add Player Detection Handler.
- Change Menu UI.
- Add Options Button and Back To Menu button in-game.

## Version 0.1.5.1
- This version is compatible with Server v0.1.5
- Adding gamepad controller (Beta)

## Version 0.1.5
- Skip versioning so it's the same as server.
- Improve bandwidth usage, using formatted message instead of CSV.
- Embed Chat add background color.

## Version 0.1
- First Release

scripts/three/modules/Stage.js
import {
    BoxBufferGeometry,
    Math,
    MeshStandardMaterial,
    MeshBasicMaterial,
    Mesh,
    TextureLoader,
    PlaneBufferGeometry,
    PointLight
} from '../three.module.js';

import { BufferGeometryUtils } from "./BuffersGeometryUtils.js";

import { Rabbit } from './Rabbit.js';
import { Speaker } from './Speaker.js';

var Stage = function(rabbit, scene) {

    // Stage is Ready?
    this.readyCount = 0;

    this.scene = scene;
    this.rabbit = rabbit;
    this.showAll = true;

    var stageGeo = [];

    var stageTop = new BoxBufferGeometry( 5000, 300, 500 );
    var stageBottom = new BoxBufferGeometry( 5000, 300, 840 );
    stageTop.translate(0, 840, 0);
    stageBottom.translate(0, -840, -170);

    var stageBack = new BoxBufferGeometry( 3400, 1380, 100 );
    stageBack.translate(0, 0, 300);

    var stageLeft = new BoxBufferGeometry( 300, 1380, 500 );
    var stageRight = stageLeft.clone();
    stageLeft.translate(1850, 0, 0);
    stageRight.translate(-1850, 0, 0);

    stageGeo.push(stageBottom);
    stageGeo.push(stageTop);
    stageGeo.push(stageBack);
    stageGeo.push(stageLeft);
    stageGeo.push(stageRight);

    var stageMerge = BufferGeometryUtils.mergeBufferGeometries(stageGeo);

    var stageMaterial = new MeshStandardMaterial( {
        emissive: 0x080808,
        color: 0x141414,
        roughness: 0.3,
        metalness: 0,
        transparent: true,
        opacity: 0.7
    });

    this.stage = new Mesh( stageMerge, stageMaterial );
    this.stage.position.set(2000, 690, 3600);

    this.scene.add(this.stage);

    // Create Lights
    this.backLightDelay = 1500;
    this.backLightFade = 1250;
    this.backLightDelta = 0;
    this.backLightState = 0; // 0 to on, 1 to off, 2 always on, 3 always off
    this.backLight = new PointLight(0x4857b5, 10, 4000, 1);
    this.backLight.position.set(2000, 690, 3700);
    this.scene.add(this.backLight);

    // Create Speaker
    this.speakerLeft = new Speaker();
    this.speakerLeft.position.set(this.stage.position.x + 1330, 765, this.stage.position.z);
    this.speakerLeft.lookAt(this.speakerLeft.position.x, this.speakerLeft.position.y, this.speakerLeft.position.z - 10);

    this.speakerRight = this.speakerLeft.clone();
    this.speakerRight.position.set(this.stage.position.x - 1330, 765, this.stage.position.z);
    this.speakerRight.lookAt(this.speakerRight.position.x, this.speakerRight.position.y, this.speakerRight.position.z - 10);

    this.scene.add(this.speakerLeft);
    this.scene.add(this.speakerRight);

    // Create TV Box
    var tvBoxGeo = [];
    var tvBox = new BoxBufferGeometry( 2020, 1180, 50 );
    var tvBoxBorderTop = new BoxBufferGeometry( 2020, 30, 10 );
    var tvBoxBorderBottom = tvBoxBorderTop.clone();
    tvBoxBorderTop.translate(0, 575, 30);
    tvBoxBorderBottom.translate(0, -575, 30);

    var tvBoxBorderLeft = new BoxBufferGeometry( 30, 1180, 10 );
    var tvBoxBorderRight = new BoxBufferGeometry( 30, 1180, 10 );
    tvBoxBorderLeft.translate(-995, 0, 30);
    tvBoxBorderRight.translate(995, 0, 30);

    var tvBoxBack = new BoxBufferGeometry( 1420, 680, 50 );
    tvBoxBack.translate(0, 0, -50);

    var tvBoxStick = new BoxBufferGeometry( 200, 690, 50 );
    tvBoxStick.translate(0, -345, -100);

    var tvBoxBottom = new BoxBufferGeometry( 700, 20, 200 );
    tvBoxBottom.translate(0, -680, -75);

    tvBoxGeo.push(tvBox);
    tvBoxGeo.push(tvBoxBorderTop);
    tvBoxGeo.push(tvBoxBorderBottom);
    tvBoxGeo.push(tvBoxBorderLeft);
    tvBoxGeo.push(tvBoxBorderRight);
    tvBoxGeo.push(tvBoxBack);
    tvBoxGeo.push(tvBoxStick);
    tvBoxGeo.push(tvBoxBottom);

    var tvBoxMerge = BufferGeometryUtils.mergeBufferGeometries(tvBoxGeo);

    var tvBoxMaterial = new MeshStandardMaterial( {
        emissive: 0x0,
        color: 0x141414,
        roughness: 0.1,
        metalness: 0
    });

    this.tvBox = new Mesh( tvBoxMerge, tvBoxMaterial );
    this.tvBox.position.set(this.stage.position.x, 690, this.stage.position.z + 25);
    this.tvBox.lookAt(this.tvBox.position.x, this.tvBox.position.y, this.tvBox.position.z - 10);

    this.scene.add(this.tvBox);

    // Create Big Black Rabbit Stage
    this.bigRabbit = new Rabbit(0, 0, 10, 205, 205, 205);
    this.bigRabbit.object.scale.set(50, 50, 50);
    this.bigRabbit.object.position.set(this.stage.position.x, 2200, this.stage.position.z + 300);
    this.bigRabbit.object.lookAt(0, this.bigRabbit.object.position.y, 0);
    this.bigRabbit.object.material.transparent = true;
    this.bigRabbit.object.material.opacity = 0.7;

    this.scene.add(this.bigRabbit.object);

    this.smallRabbit = new Rabbit(0, 0, 50);
    this.smallRabbit.object.scale.set(3, 3, 3);
    this.smallRabbit.object.position.set(this.stage.position.x - 380, 35, this.stage.position.z + 50);
    this.smallRabbit.object.lookAt(this.smallRabbit.object.position.x, this.smallRabbit.object.position.y, this.smallRabbit.object.position.z - 10);
    this.smallRabbit.object.rotation.z = Math.degToRad(155);

    this.scene.add(this.smallRabbit.object);

    // Image stuff!
    var that = this;

    var textureLoader = new TextureLoader();

    // Spinning Logo
    this.logoTop;
    this.blockLogoLeft;
    this.blockLogoRight;
    this.bannerLeft;
    this.bannerRight;

    textureLoader.load(
        'images/stage/logo.png',
        function (texture) {
            var material = new MeshBasicMaterial( { map: texture, transparent: true } );
            var geometry = new PlaneBufferGeometry( 4950, 250 );
            that.logoTop = new Mesh( geometry, material );
            that.logoTop.position.set(that.stage.position.x, 1530, that.stage.position.z - 252)
            that.logoTop.lookAt(that.logoTop.position.x, that.logoTop.position.y, 0);
            that.logoTop.renderOrder = 1;
            if(that.showAll) {
                that.scene.add(that.logoTop);
            }
            that.readyCount++;
        },
        undefined,
	    function ( err ) {
            console.error("No Logo Top Found.");
            that.readyCount++;
        }
    );

    textureLoader.load(
        'images/stage/logo-box.png',
        function (texture) {
            var blockLogoGeo = new BoxBufferGeometry( 200, 200, 200 );
            var blockLogoMaterial = new MeshStandardMaterial( {
                map: texture,
                transparent: true,
                roughness: 0.5,
                metalness: 0
            });
            that.blockLogoLeft = new Mesh( blockLogoGeo, blockLogoMaterial );
            that.blockLogoRight = that.blockLogoLeft.clone();
            that.blockLogoLeft.position.set(that.stage.position.x + 1330, 1000, that.stage.position.z);
            that.blockLogoRight.position.set(that.stage.position.x - 1330, 1000, that.stage.position.z);
            that.blockLogoLeft.renderOrder = 1;
            that.blockLogoRight.renderOrder = 1;
            if(that.showAll) {
                that.scene.add(that.blockLogoLeft);
                that.scene.add(that.blockLogoRight);
            }
            that.readyCount++;
        },
        undefined,
	    function ( err ) {
            console.error("No Logo Box Found.");
            that.readyCount++;
        }
    );


    textureLoader.load(
        'images/stage/banner-left.png',
        function (texture) {
            var material = new MeshBasicMaterial( { map: texture, transparent: true } );
            var geometry = new PlaneBufferGeometry( 250, 1280 );
            that.bannerLeft = new Mesh( geometry, material );
            that.bannerLeft.position.set(that.stage.position.x + 1850, 690, that.stage.position.z - 255)
            that.bannerLeft.lookAt(that.bannerLeft.position.x, that.bannerLeft.position.y, 0);
            that.bannerLeft.renderOrder = 1;
            if(that.showAll) {
                that.scene.add(that.bannerLeft);
            }
            that.readyCount++;
        },
        undefined,
	    function ( err ) {
            console.error("No Banner Left Found.");
            that.readyCount++;
        }
    );

    textureLoader.load(
        'images/stage/banner-right.png',
        function (texture) {
            var material = new MeshBasicMaterial( { map: texture, transparent: true } );
            var geometry = new PlaneBufferGeometry( 250, 1280 );
            that.bannerRight = new Mesh( geometry, material );
            that.bannerRight.position.set(that.stage.position.x - 1850, 690, that.stage.position.z - 255)
            that.bannerRight.lookAt(that.bannerRight.position.x, that.bannerRight.position.y, 0);
            that.bannerRight.renderOrder = 1;
            if(that.showAll) {
                that.scene.add(that.bannerRight);
            }
            that.readyCount++;
        },
        undefined,
	    function ( err ) {
            console.error("No Banner Right Found.");
            that.readyCount++;
        }
    );

    this.isReady = function() {
        if(this.readyCount >= 4) {
            return true;
        }

        return false;
    };

    this.lightAlwaysOff = function() {
        this.backLightState = 3;
    };

    this.lightAlwaysOn = function() {
        this.backLightState = 2;
    };

    this.lightBlink = function(delay, fade) {
        if(this.backLightState > 1) {
            this.backLightState = 0;
            this.backLightDelta = 0;
        }
        this.backLightDelay = delay;
        this.backLightFade = fade;
    };

    this.lightColor = function(color) {
        this.backLight.color.setStyle(color);
    };

    this.lightDisable = function() {
        this.backLight.visible = false;
    };

    this.lightEnable = function() {
        this.backLight.visible = true;
    };

    this.update = function(delta) {
        if(this.hideAll) {
            return;
        }

        if(this.backLight.visible) {
            if(this.backLightState == 0) {
                this.backLightDelta += (delta * 1000);
                if(this.backLightDelta > this.backLightDelay) {
                    this.backLightDelta = this.backLightDelta % this.backLightDelay;
                    this.backLightState = 1;
                    this.backLight.intensity = 10 - (10 * ((this.backLightDelta%this.backLightFade) / this.backLightFade));
                } else {
                    if(this.backLightDelta < this.backLightFade) {
                        this.backLight.intensity = 10 * (this.backLightDelta / this.backLightFade);
                    }
                }
            } else if(this.backLightState == 1) {
                    this.backLightDelta += (delta * 1000);
                    if(this.backLightDelta > this.backLightDelay) {
                        this.backLightDelta = this.backLightDelta % this.backLightDelay;
                        this.backLightState = 0;
                        this.backLight.intensity = 10 * ((this.backLightDelta%this.backLightFade) / this.backLightFade);
                    } else {
                        if(this.backLightDelta < this.backLightFade) {
                            this.backLight.intensity = 10 - (10 * (this.backLightDelta / this.backLightFade));
                        }
                    }
            } else if(this.backLightState == 2 && this.backLight.intensity != 10) {
                this.backLight.intensity = 10;
            } else if(this.backLightState == 3 && this.backLight.intensity != 0) {
                this.backLight.intensity = 0;
            }
        }

        this.bigRabbit.object.lookAt(rabbit.object.position.x, this.bigRabbit.object.position.y, rabbit.object.position.z);
        if(typeof this.blockLogoLeft !== 'undefined' && typeof this.blockLogoRight !== 'undefined') {
            this.blockLogoLeft.rotation.y -= delta * 0.5;
            this.blockLogoRight.rotation.y += delta * 0.5;
        }
    };

    this.hide = function(){
        if(this.showAll){
            this.showAll = false;
            this.scene.remove(this.stage);
            this.scene.remove(this.speakerLeft);
            this.scene.remove(this.speakerRight);
            this.scene.remove(this.tvBox);
            this.scene.remove(this.bigRabbit.object);
            this.scene.remove(this.smallRabbit.object);

            if(this.isReady()) {
                this.scene.remove(this.blockLogoLeft);
                this.scene.remove(this.blockLogoRight);
                this.scene.remove(this.logoTop);
                this.scene.remove(this.bannerLeft);
                this.scene.remove(this.bannerRight);
            }
        }
    };

    this.show = function(){
        if(!this.showAll){
            this.showAll = true;
            this.scene.add(this.stage);
            this.scene.add(this.speakerLeft);
            this.scene.add(this.speakerRight);
            this.scene.add(this.tvBox);
            this.scene.add(this.bigRabbit.object);
            this.scene.add(this.smallRabbit.object);

            if(this.isReady()) {
                this.scene.add(this.blockLogoLeft);
                this.scene.add(this.blockLogoRight);
                this.scene.add(this.logoTop);
                this.scene.add(this.bannerLeft);
                this.scene.add(this.bannerRight);
            }
        }
    };

};

export { Stage };